# Shrek sentimental project

It will be continued working with the csv table that was worked previously. Table with translated column now is ready to start the sentiment analysis between reaction of IG comments

In [11]:
# Importation tools of the development of the project
import pandas as pd
import numpy as np
import seaborn as sns
import nltk
import emoji
from tqdm import tqdm
from nltk.sentiment import SentimentIntensityAnalyzer
from tqdm.notebook import tqdm
sia = SentimentIntensityAnalyzer()


nltk.download('punkt')
nltk.download('averaged_perceptron_tagger_eng')
nltk.download('maxent_ne_chunker_tab')
nltk.download('words')
nltk.download('vader_lexicon')
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package punkt to C:\Users\hdbl2/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     C:\Users\hdbl2/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package maxent_ne_chunker_tab to
[nltk_data]     C:\Users\hdbl2/nltk_data...
[nltk_data]   Package maxent_ne_chunker_tab is already up-to-date!
[nltk_data] Downloading package words to C:\Users\hdbl2/nltk_data...
[nltk_data]   Package words is already up-to-date!
[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\hdbl2/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [12]:
# data importation from instagram Srek 5 post scrapping
df = pd.read_csv('C:\\Users\hdbl2\OneDrive\Desktop\CursoPython\Datsets\\shrek_commets_INST.csv', delimiter=';')

## 1- Data Cleaning and Filtering
Lets take a look of the orignal dataset

In [13]:
# Previsualization of the original data
df

comment_id    created_at  \
0     18108918547480299  1.741234e+09   
1     17922193934943166  1.741235e+09   
2     18390414847129171  1.741235e+09   
3     18022853681435244  1.741235e+09   
4     18038712758432147  1.741235e+09   
...                 ...           ...   
9920  17898548340144291  1.740838e+09   
9921  18057937055021390  1.740838e+09   
9922  18029147957334468  1.740838e+09   
9923  18051238781329295  1.740838e+09   
9924  18219148771292675  1.740838e+09   

                                        profile_pic_url  \
0     https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
1     https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
2     https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
3     https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
4     https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
...                                                 ...   
9920  https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
9921  https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
9922  https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
9923  https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   
9924  https://instagram.fgdl10-1.fna.fbcdn.net/v/t51...   

                                                   text       user_id  \
0                                            Muito ruim  1.774686e+09   
1                                                  Why😭  6.999667e+10   
2                                                     😭  7.170031e+10   
3     Can you restore Shrek's face to the previous v...  1.160897e+10   
4                                                  🖕👎🖕👎  1.160897e+10   
...                                                 ...           ...   
9920                                              💔💔💔💔💔  1.805980e+09   
9921                                                NaN  6.891905e+10   
9922                                        #notmyshrek  1.408508e+10   
9923                                                 🗿🫷  3.254513e+10   
9924                      Zendaya again???????? Why????  4.728614e+09   

               username  
0            pablobrumm  
1     hooman_saghi.2010  
2       bensajes_nahual  
3     hiking_with_saeed  
4     hiking_with_saeed  
...                 ...  
9920  karol_goncalves_m  
9921        bryant20081  
9922         chqrlottek  
9923        _matador_ox  
9924      julio_silbass  

[9925 rows x 6 columns]

## 2- Exploratory Data Analysis (EDA) 
We can see that they´re (9925) rows of data, and some columns that are irrelevant for our analysis. In the the Transltaion_of_comments code, it was realized some important aspects like:

#
- Clean our data in ["text"] column section, removing the null values. After cleaning that column, the dataset was reduced to (8712) rows of data.

#
- Filter the the neccesary column for the analysis, excluiding columns as ["profile_pic_url"] and the ["user_id"].

#
Note: Excluding the ["profile_pic_url"] and the ["user_id"], is for respect the privacy for the users, and furthermore, the questions to be answered do not cover relationship about the user in general.

## 3- Insights and Queries

In [14]:
#Importation of the table with comments translated was made in the Translation_of_comments code
table_transl = pd.read_csv("Table_translated.csv")

In [15]:
#Previsualization of the data
table_transl

comment_id           username  \
0     18108918547480299         pablobrumm   
1     17922193934943166  hooman_saghi.2010   
2     18390414847129171    bensajes_nahual   
3     18022853681435244  hiking_with_saeed   
4     18038712758432147  hiking_with_saeed   
...                 ...                ...   
8707  18059369398978547      lucas_ra10.02   
8708  17898548340144291  karol_goncalves_m   
8709  18029147957334468         chqrlottek   
8710  18051238781329295        _matador_ox   
8711  18219148771292675      julio_silbass   

                                                   text  \
0                                            Muito ruim   
1                                                  Why😭   
2                                                     😭   
3     Can you restore Shrek's face to the previous v...   
4                                                  🖕👎🖕👎   
...                                                 ...   
8707                                     @habi_65_ aqui   
8708                                              💔💔💔💔💔   
8709                                        #notmyshrek   
8710                                                 🗿🫷   
8711                      Zendaya again???????? Why????   

                                    translated_comments    created_at  
0                                              Very bad  1.741234e+09  
1                                                  Why😭  1.741235e+09  
2                                                   NaN  1.741235e+09  
3     Can you restore Shrek's face to the previous v...  1.741235e+09  
4                                                   NaN  1.741235e+09  
...                                                 ...           ...  
8707                                    @inhab_65_ here  1.740838e+09  
8708                                                NaN  1.740838e+09  
8709                                        #NOTMYSHREK  1.740838e+09  
8710                                                NaN  1.740838e+09  
8711                      Zendaya again???????? Why????  1.740838e+09  

[8712 rows x 5 columns]

In [16]:
Shrek_dataset_TR = table_transl

In [17]:
#Convert the scientific notation values of the creation of the comment to dates that we can contamplate 
import datetime
Shrek_dataset_TR['created_at'] = Shrek_dataset_TR['created_at'].apply(lambda x: datetime.datetime.utcfromtimestamp(x))

In [18]:
#Previsualization of the table
Shrek_dataset_TR.head()

comment_id           username  \
0  18108918547480299         pablobrumm   
1  17922193934943166  hooman_saghi.2010   
2  18390414847129171    bensajes_nahual   
3  18022853681435244  hiking_with_saeed   
4  18038712758432147  hiking_with_saeed   

                                                text  \
0                                         Muito ruim   
1                                               Why😭   
2                                                  😭   
3  Can you restore Shrek's face to the previous v...   
4                                               🖕👎🖕👎   

                                 translated_comments          created_at  
0                                           Very bad 2025-03-06 04:09:21  
1                                               Why😭 2025-03-06 04:16:00  
2                                                NaN 2025-03-06 04:17:50  
3  Can you restore Shrek's face to the previous v... 2025-03-06 04:24:33  
4                                                NaN 2025-03-06 04:24:49

### 3.1- Sentiment Analysis by text and emojis

In [19]:
#Using VADER library to analyze the text of the commentd
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

# Manually add some words to VADER’s lexicon and improve the model
new_words = {
    "disapprove": -2.0, "thumbs down": -2.0, "bad emoji": -2.0,
    "laughing": 2.0, "love": 2.0, "crying": -1.5, "angry": -2.0, "dead": -1.5
}
sia.lexicon.update(new_words)

# Dictionary to detect the sentiment of the emoji
emoji_sentiment_dict = {
    "😂": "Positive", "🤣": "Positive", "😭": "Negative", "😊": "Positive", "❤️": "Positive",
    "😢": "Negative", "😡": "Negative", "😠": "Negative", "💔": "Negative",
    "😐": "Neutral", "🖕": "Negative", "👎": "Negative", "💀": "Negative", "🤮": "Negative", "🤢": "Negative"
}

# Function to analyze emoji sentiment separately using the tex column
#that was the original column, rather than the translated column to be applied below

def analyze_emoji_sentiment(text):
    if pd.isna(text) or not isinstance(text, str):  
        return "No Emojis", 0  #Is there no emojis return the value

    # Extract only emojis
    emojis = ''.join(char for char in text if emoji.is_emoji(char))
    
    if not emojis:
        return "No Emojis", 0  

    # Check sentiment of each emoji with our previous dictionary
    sentiments = [emoji_sentiment_dict.get(char, "Neutral") for char in emojis]
    
    if "Negative" in sentiments:
        return "Negative", -0.6  # NEGATIVE WEIGHT
    elif "Positive" in sentiments:
        return "Positive", 0.6   # POSITIVE WEIGHT
    else:
        return "Neutral", 0  

#Dictionary to store the results
results = {}

#More impact in emojis: Visualizing the data, was found some comments
#with a neutral text and a negative emoji, the model recognized as a neutral comment,
#to fix this issue, it was incremented their valu to consider it in the final score
EMOJI_WEIGHT = 1.5  

# Application of the sentiment analysis of the text using VADER, row by row, in the translated column
for i, row in tqdm(Shrek_dataset_TR.iterrows(), total=len(Shrek_dataset_TR)):
    user = row["username"]  # Identify user
    translated_text = str(row["translated_comments"]).strip()

    # VADER sentiment on translated comments
    text_sentiment = sia.polarity_scores(translated_text) if translated_text else {"compound": 0, "neg": 0, "neu": 1, "pos": 0}
    
    # Application of the Emoji sentiment analysis on original comments "text" column
    emoji_text = str(row["text"]).strip()
    emoji_sentiment, emoji_adjustment = analyze_emoji_sentiment(emoji_text) #Use of emojis function for analysis

    # Final compound score with increased the emoji weight for improve the analysis
    final_compound = text_sentiment["compound"] + (emoji_adjustment * EMOJI_WEIGHT)
    final_compound = max(-1, min(1, final_compound)) 

    # Store results in the dictionary
    results[user] = {
        "text_sentiment": text_sentiment["compound"],  # Store only compound score for clarity
        "emoji_sentiment": emoji_sentiment,
        "final_sentiment": final_compound
    }

# Convert the results dictionary into a DataFrame
df_sentiment = pd.DataFrame.from_dict(results, orient="index").reset_index().rename(columns={"index": "username"})

# Merge results back into the previous dataset
Shrek_dataset_TR = Shrek_dataset_TR.merge(df_sentiment, on="username", how="left")


  0%|          | 0/8712 [00:00<?, ?it/s]

In [20]:
#Previsualization of the dataset
Shrek_dataset_TR

comment_id           username  \
0     18108918547480299         pablobrumm   
1     17922193934943166  hooman_saghi.2010   
2     18390414847129171    bensajes_nahual   
3     18022853681435244  hiking_with_saeed   
4     18038712758432147  hiking_with_saeed   
...                 ...                ...   
8707  18059369398978547      lucas_ra10.02   
8708  17898548340144291  karol_goncalves_m   
8709  18029147957334468         chqrlottek   
8710  18051238781329295        _matador_ox   
8711  18219148771292675      julio_silbass   

                                                   text  \
0                                            Muito ruim   
1                                                  Why😭   
2                                                     😭   
3     Can you restore Shrek's face to the previous v...   
4                                                  🖕👎🖕👎   
...                                                 ...   
8707                                     @habi_65_ aqui   
8708                                              💔💔💔💔💔   
8709                                        #notmyshrek   
8710                                                 🗿🫷   
8711                      Zendaya again???????? Why????   

                                    translated_comments          created_at  \
0                                              Very bad 2025-03-06 04:09:21   
1                                                  Why😭 2025-03-06 04:16:00   
2                                                   NaN 2025-03-06 04:17:50   
3     Can you restore Shrek's face to the previous v... 2025-03-06 04:24:33   
4                                                   NaN 2025-03-06 04:24:49   
...                                                 ...                 ...   
8707                                    @inhab_65_ here 2025-03-01 14:05:13   
8708                                                NaN 2025-03-01 14:05:15   
8709                                        #NOTMYSHREK 2025-03-01 14:05:38   
8710                                                NaN 2025-03-01 14:05:45   
8711                      Zendaya again???????? Why???? 2025-03-01 14:05:58   

      text_sentiment emoji_sentiment  final_sentiment  
0            -0.5849       No Emojis          -0.5849  
1             0.0000        Negative          -0.9000  
2             0.0000        Negative          -0.9000  
3             0.0000        Negative          -0.9000  
4             0.0000        Negative          -0.9000  
...              ...             ...              ...  
8707          0.0000       No Emojis           0.0000  
8708          0.0000        Negative          -0.9000  
8709          0.0000       No Emojis           0.0000  
8710          0.0000         Neutral           0.0000  
8711          0.0000       No Emojis           0.0000  

[8712 rows x 8 columns]

### 3.2- Detection of the region by their langauge

In [21]:
#Application of the assignation of region by detecting the language 
#Library for detecting the language
import langid

#Function that checks if text column contains only emojis
def is_only_emojis(text):
    return isinstance(text, str) and all(emoji.is_emoji(char) for char in text.strip())

#Function to detect language safely using langid
def detect_language(text):
    if is_only_emojis(text):  
        return "emoji"  #Assign a special label identificatioj for only emoji comments
    try:
        return langid.classify(text)[0]  # Detect language
    except:
        return "unknown"  # If detection fails, return "unknown"

# Function to map detected language to a general region
def map_language_to_region(lang):
    language_region_map = {
        "en": "North America / UK",
        "es": "Latin America / Spain",
        "fr": "France / Canada",
        "de": "Germany / Central Europe",
        "pt": "Brazil / Portugal",
        "it": "Italy",
        "ru": "Russia / Eastern Europe",
        "zh": "China",
        "ja": "Japan",
        "ko": "Korea",
        "ar": "Middle East",
        "hi": "India",
        "nl": "Netherlands",
        "unknown": "Other",
        "emoji": "Emoji-Based Region"  #Custom region for emoji-only comments
    }
    return language_region_map.get(lang, "Other")

#Apply `langid` language detection to the "text" column
Shrek_dataset_TR["detected_language"] = Shrek_dataset_TR["text"].apply(detect_language)

# Map detected languages to regions
Shrek_dataset_TR["region"] = Shrek_dataset_TR["detected_language"].apply(map_language_to_region)

In [22]:
Shrek_dataset_TR

comment_id           username  \
0     18108918547480299         pablobrumm   
1     17922193934943166  hooman_saghi.2010   
2     18390414847129171    bensajes_nahual   
3     18022853681435244  hiking_with_saeed   
4     18038712758432147  hiking_with_saeed   
...                 ...                ...   
8707  18059369398978547      lucas_ra10.02   
8708  17898548340144291  karol_goncalves_m   
8709  18029147957334468         chqrlottek   
8710  18051238781329295        _matador_ox   
8711  18219148771292675      julio_silbass   

                                                   text  \
0                                            Muito ruim   
1                                                  Why😭   
2                                                     😭   
3     Can you restore Shrek's face to the previous v...   
4                                                  🖕👎🖕👎   
...                                                 ...   
8707                                     @habi_65_ aqui   
8708                                              💔💔💔💔💔   
8709                                        #notmyshrek   
8710                                                 🗿🫷   
8711                      Zendaya again???????? Why????   

                                    translated_comments          created_at  \
0                                              Very bad 2025-03-06 04:09:21   
1                                                  Why😭 2025-03-06 04:16:00   
2                                                   NaN 2025-03-06 04:17:50   
3     Can you restore Shrek's face to the previous v... 2025-03-06 04:24:33   
4                                                   NaN 2025-03-06 04:24:49   
...                                                 ...                 ...   
8707                                    @inhab_65_ here 2025-03-01 14:05:13   
8708                                                NaN 2025-03-01 14:05:15   
8709                                        #NOTMYSHREK 2025-03-01 14:05:38   
8710                                                NaN 2025-03-01 14:05:45   
8711                      Zendaya again???????? Why???? 2025-03-01 14:05:58   

      text_sentiment emoji_sentiment  final_sentiment detected_language  \
0            -0.5849       No Emojis          -0.5849                it   
1             0.0000        Negative          -0.9000                en   
2             0.0000        Negative          -0.9000             emoji   
3             0.0000        Negative          -0.9000                en   
4             0.0000        Negative          -0.9000             emoji   
...              ...             ...              ...               ...   
8707          0.0000       No Emojis           0.0000                ca   
8708          0.0000        Negative          -0.9000             emoji   
8709          0.0000       No Emojis           0.0000                en   
8710          0.0000         Neutral           0.0000             emoji   
8711          0.0000       No Emojis           0.0000                en   

                  region  
0                  Italy  
1     North America / UK  
2     Emoji-Based Region  
3     North America / UK  
4     Emoji-Based Region  
...                  ...  
8707               Other  
8708  Emoji-Based Region  
8709  North America / UK  
8710  Emoji-Based Region  
8711  North America / UK  

[8712 rows x 10 columns]

In [24]:
 Shrek_dataset_TR = Shrek_dataset_TR[["comment_id", "username", "text", "translated_comments", "text_sentiment", "emoji_sentiment", "final_sentiment", "detected_language", "region", "created_at"]]

In [25]:
Shrek_dataset_TR

comment_id           username  \
0     18108918547480299         pablobrumm   
1     17922193934943166  hooman_saghi.2010   
2     18390414847129171    bensajes_nahual   
3     18022853681435244  hiking_with_saeed   
4     18038712758432147  hiking_with_saeed   
...                 ...                ...   
8707  18059369398978547      lucas_ra10.02   
8708  17898548340144291  karol_goncalves_m   
8709  18029147957334468         chqrlottek   
8710  18051238781329295        _matador_ox   
8711  18219148771292675      julio_silbass   

                                                   text  \
0                                            Muito ruim   
1                                                  Why😭   
2                                                     😭   
3     Can you restore Shrek's face to the previous v...   
4                                                  🖕👎🖕👎   
...                                                 ...   
8707                                     @habi_65_ aqui   
8708                                              💔💔💔💔💔   
8709                                        #notmyshrek   
8710                                                 🗿🫷   
8711                      Zendaya again???????? Why????   

                                    translated_comments  text_sentiment  \
0                                              Very bad         -0.5849   
1                                                  Why😭          0.0000   
2                                                   NaN          0.0000   
3     Can you restore Shrek's face to the previous v...          0.0000   
4                                                   NaN          0.0000   
...                                                 ...             ...   
8707                                    @inhab_65_ here          0.0000   
8708                                                NaN          0.0000   
8709                                        #NOTMYSHREK          0.0000   
8710                                                NaN          0.0000   
8711                      Zendaya again???????? Why????          0.0000   

     emoji_sentiment  final_sentiment detected_language              region  \
0          No Emojis          -0.5849                it               Italy   
1           Negative          -0.9000                en  North America / UK   
2           Negative          -0.9000             emoji  Emoji-Based Region   
3           Negative          -0.9000                en  North America / UK   
4           Negative          -0.9000             emoji  Emoji-Based Region   
...              ...              ...               ...                 ...   
8707       No Emojis           0.0000                ca               Other   
8708        Negative          -0.9000             emoji  Emoji-Based Region   
8709       No Emojis           0.0000                en  North America / UK   
8710         Neutral           0.0000             emoji  Emoji-Based Region   
8711       No Emojis           0.0000                en  North America / UK   

              created_at  
0    2025-03-06 04:09:21  
1    2025-03-06 04:16:00  
2    2025-03-06 04:17:50  
3    2025-03-06 04:24:33  
4    2025-03-06 04:24:49  
...                  ...  
8707 2025-03-01 14:05:13  
8708 2025-03-01 14:05:15  
8709 2025-03-01 14:05:38  
8710 2025-03-01 14:05:45  
8711 2025-03-01 14:05:58  

[8712 rows x 10 columns]

## Load of the final dataset

In [26]:
Shrek_dataset_TR.to_csv("Final_Shrek_dataset.csv", index=False)